pkg/store/container/container.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package container
+
+import (
+	"sync"
+
+	"github.com/kubernetes-incubator/cri-containerd/pkg/store"
+)
+
+// Container contains all resources associated with the container. All fields of
+// Container are thread safe.
+type Container struct {
+	// Metadata is the metadata of the container, it is **read-only**.
+	Metadata
+	// Status stores the status of the container.
+	Status StatusStorage
+	// TODO(random-liu): Add containerd container client.
+	// TODO(random-liu): Add stop channel to get rid of stop poll waiting.
+}
+
+// NewContainer creates a internal used container type.
+func NewContainer(metadata Metadata, status Status) (Container, error) {
+	s, err := StoreStatus(metadata.ID, status)
+	if err != nil {
+		return Container{}, err
+	}
+	return Container{
+		Metadata: metadata,
+		Status:   s,
+	}, nil
+}
+
+// LoadContainer loads the internal used container type.
+func LoadContainer() (Container, error) {
+	return Container{}, nil
+}
+
+// Store stores all Containers.
+type Store struct {
+	lock       sync.RWMutex
+	containers map[string]Container
+	// TODO(random-liu): Add trunc index.
+}
+
+// LoadStore loads containers from runtime.
+// TODO(random-liu): Implement LoadStore.
+func LoadStore() *Store { return nil }
+
+// NewStore creates a container store.
+func NewStore() *Store {
+	return &Store{containers: make(map[string]Container)}
```

Comment:
I use `Container` rather than `*Container` because:
1) The value copy could prevent callers from mutating the struct fields in some way, because most fields are primitive types.
2) Most importantly, when caller using the api, usually they won't expect to be able to mutate fields of a returned value, but may expect to mutate a returned pointer.

As for the overhead, metadata is small, so there shouldn't be too much overhead. :)